File: app/src/main/java/info/nightscout/utils/FabricPrivacy.java
package info.nightscout.utils;

import com.crashlytics.android.Crashlytics;
import com.crashlytics.android.answers.Answers;
import com.crashlytics.android.answers.CustomEvent;

import info.nightscout.androidaps.BuildConfig;
import info.nightscout.androidaps.Config;
import info.nightscout.androidaps.MainApp;
import info.nightscout.androidaps.R;
import info.nightscout.androidaps.interfaces.PluginBase;

import java.util.Date;

/**
 * Created by jamorham on 21/02/2018.
 * <p>
 * Some users do not wish to be tracked, Fabric Answers and Crashlytics do not provide an easy way
 * to disable them and make calls from a potentially invalid singleton reference. This wrapper
 * emulates the methods but ignores the request if the instance is null or invalid.
 */

public class FabricPrivacy {

    private static final String TAG = "FabricPrivacy";
    private static volatile FabricPrivacy instance;


    public static FabricPrivacy getInstance() {
        if (instance == null) {
            initSelf();
        }
        return instance;
    }

    private static synchronized void initSelf() {
        if (instance == null) {
            instance = new FabricPrivacy();
        }
    }

    // Crashlytics logException
    public static void logException(Throwable throwable) {
        try {
            final Crashlytics crashlytics = Crashlytics.getInstance();
            crashlytics.core.logException(throwable);
        } catch (NullPointerException | IllegalStateException e) {
            android.util.Log.d(TAG, "Ignoring opted out non-initialized log: " + throwable);
        }
    }

    // Crashlytics log
    public static void log(String msg) {
        try {
            final Crashlytics crashlytics = Crashlytics.getInstance();
            crashlytics.core.log(msg);
        } catch (NullPointerException | IllegalStateException e) {
            android.util.Log.d(TAG, "Ignoring opted out non-initialized log: " + msg);
        }
    }

    // Crashlytics log
    public static void log(int priority, String tag, String msg) {
        try {
            final Crashlytics crashlytics = Crashlytics.getInstance();
            crashlytics.core.log(priority, tag, msg);
        } catch (NullPointerException | IllegalStateException e) {
            android.util.Log.d(TAG, "Ignoring opted out non-initialized log: " + msg);
        }
    }

    public static boolean fabricEnabled() {
        return SP.getBoolean("enable_fabric", true);
    }

    // Answers logCustom
    public void logCustom(CustomEvent event) {
        try {
            final Answers answers = Answers.getInstance();
            if (fabricEnabled()) {
                answers.logCustom(event);
            } else {
                android.util.Log.d(TAG, "Ignoring recently opted-out event: " + event.toString());
            }
        } catch (NullPointerException | IllegalStateException e) {
            android.util.Log.d(TAG, "Ignoring opted-out non-initialized event: " + event.toString());
        }
    }

    public static void uploadDailyStats() {
        if (!fabricEnabled()) return;

        long lastUploadDay = SP.getLong(MainApp.gs(R.string.key_plugin_stats_report_timestamp), 0L);

        Date date = new Date();
        date.setHours(0);
        date.setMinutes(0);
        date.setSeconds(0);
        long today = date.getTime() - date.getTime() % 1000;

        if (today > lastUploadDay) {
            uploadAppUsageType();
            uploadPluginStats();

            SP.putLong(MainApp.gs(R.string.key_plugin_stats_report_timestamp), today);
        }
    }

    private static void uploadPluginStats() {
        CustomEvent pluginStats = new CustomEvent("PluginStats");
        pluginStats.putCustomAttribute("version", BuildConfig.VERSION);
        pluginStats.putCustomAttribute("HEAD", BuildConfig.HEAD);
        pluginStats.putCustomAttribute("language", SP.getString(R.string.key_language,"default"));
        for (PluginBase plugin : MainApp.getPluginsList()) {
            if (plugin.isEnabled(plugin.getType()) && !plugin.pluginDescription.alwaysEnabled) {
                // Fabric allows no more than 20 attributes attached to an event. By reporting disabled plugins as
                // well, we would exceed that threshold, so only report what is enabled
                // TODO >2.0: consider reworking this to upload an event per enabled plugin instead.
                pluginStats.putCustomAttribute(plugin.getClass().getSimpleName(), "enabled");
            }
        }

        getInstance().logCustom(pluginStats);
    }

    private static void uploadAppUsageType() {
        CustomEvent type = new CustomEvent("AppUsageType");
        if (Config.NSCLIENT)
            type.putCustomAttribute("type", "NSClient");
        else if (Config.PUMPCONTROL)
            type.putCustomAttribute("type", "PumpControl");
        else if (MainApp.getConstraintChecker().isClosedLoopAllowed().value())
            type.putCustomAttribute("type", "ClosedLoop");
        else
            type.putCustomAttribute("type", "OpenLoop");

        getInstance().logCustom(type);
    }

}

File: core/main/src/main/java/info/nightscout/core/utils/fabric/FabricPrivacy.kt
package info.nightscout.core.utils.fabric

import android.os.Bundle
import com.google.firebase.analytics.FirebaseAnalytics
import com.google.firebase.analytics.ktx.analytics
import com.google.firebase.crashlytics.FirebaseCrashlytics
import com.google.firebase.ktx.Firebase
import info.nightscout.annotations.OpenForTesting
import info.nightscout.core.utils.R
import info.nightscout.rx.logging.AAPSLogger
import info.nightscout.rx.logging.LTag
import info.nightscout.shared.sharedPreferences.SP
import java.io.ByteArrayInputStream
import java.io.IOException
import java.io.ObjectInputStream
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Some users do not wish to be tracked, Fabric Answers and Crashlytics do not provide an easy way
 * to disable them and make calls from a potentially invalid singleton reference. This wrapper
 * emulates the methods but ignores the request if the instance is null or invalid.
 */
@OpenForTesting
@Singleton
class FabricPrivacy @Inject constructor(
    private val aapsLogger: AAPSLogger,
    private val sp: SP
) {

    val firebaseAnalytics: FirebaseAnalytics = Firebase.analytics

    init {
        firebaseAnalytics.setAnalyticsCollectionEnabled(!java.lang.Boolean.getBoolean("disableFirebase") && fabricEnabled())
        FirebaseCrashlytics.getInstance().setCrashlyticsCollectionEnabled(!java.lang.Boolean.getBoolean("disableFirebase") && fabricEnabled())
    }

    // Analytics logCustom
    @Suppress("unused")
    fun logCustom(name: String, event: Bundle) {
        try {
            if (fabricEnabled()) {
                firebaseAnalytics.logEvent(name, event)
            } else {
                aapsLogger.debug(LTag.CORE, "Ignoring recently opted-out event: $event")
            }
        } catch (e: NullPointerException) {
            aapsLogger.debug(LTag.CORE, "Ignoring opted-out non-initialized event: $event")
        } catch (e: IllegalStateException) {
            aapsLogger.debug(LTag.CORE, "Ignoring opted-out non-initialized event: $event")
        }
    }

    // Analytics logCustom
    @Suppress("unused")
    fun logCustom(event: Bundle) {
        try {
            if (fabricEnabled()) {
                firebaseAnalytics.logEvent(FirebaseAnalytics.Event.SELECT_ITEM, event)
            } else {
                aapsLogger.debug(LTag.CORE, "Ignoring recently opted-out event: $event")
            }
        } catch (e: NullPointerException) {
            aapsLogger.debug(LTag.CORE, "Ignoring opted-out non-initialized event: $event")
        } catch (e: IllegalStateException) {
            aapsLogger.debug(LTag.CORE, "Ignoring opted-out non-initialized event: $event")
        }
    }

    // Analytics logCustom
    fun logCustom(event: String) {
        try {
            if (fabricEnabled()) {
                firebaseAnalytics.logEvent(event, Bundle())
            } else {
                aapsLogger.debug(LTag.CORE, "Ignoring recently opted-out event: $event")
            }
        } catch (e: NullPointerException) {
            aapsLogger.debug(LTag.CORE, "Ignoring opted-out non-initialized event: $event")
        } catch (e: IllegalStateException) {
            aapsLogger.debug(LTag.CORE, "Ignoring opted-out non-initialized event: $event")
        }
    }

    // Crashlytics log message
    fun logMessage(message: String) {
        aapsLogger.info(LTag.CORE, "Crashlytics log message: $message")
        FirebaseCrashlytics.getInstance().log(message)
    }

    // Crashlytics logException
    fun logException(throwable: Throwable) {
        aapsLogger.error("Crashlytics log exception: ", throwable)
        FirebaseCrashlytics.getInstance().recordException(throwable)
    }

    fun fabricEnabled(): Boolean {
        return sp.getBoolean(R.string.key_enable_fabric, true)
    }

    fun logWearException(wearException: info.nightscout.rx.weardata.EventData.WearException) {
        aapsLogger.debug(LTag.WEAR, "logWearException")
        FirebaseCrashlytics.getInstance().apply {
            setCustomKey("wear_exception", true)
            setCustomKey("wear_board", wearException.board)
            setCustomKey("wear_fingerprint", wearException.fingerprint)
            setCustomKey("wear_sdk", wearException.sdk)
            setCustomKey("wear_model", wearException.model)
            setCustomKey("wear_manufacturer", wearException.manufacturer)
            setCustomKey("wear_product", wearException.product)
        }
        logException(byteArrayToThrowable(wearException.exception))
    }

    private fun byteArrayToThrowable(wearExceptionData: ByteArray): Throwable {
        val bis = ByteArrayInputStream(wearExceptionData)
        try {
            val ois = ObjectInputStream(bis)
            return ois.readObject() as Throwable
        } catch (e: IOException) {
            e.printStackTrace()
        } catch (e: ClassNotFoundException) {
            e.printStackTrace()
        }
        return IllegalArgumentException("Wear Exception could not be de-serialized")
    }
}
